front/components/agent_builder/capabilities/knowledge/CapabilitiesConfigurationSheet.tsx

```diff
@@ -0,0 +1,274 @@
+import { FormProvider } from "@app/components/sparkle/FormProvider";
+import type { DataSourceViewSelectionConfigurations } from "@app/types";
+
+interface CapabilitiesConfigurationSheetProps {
+  capability: KnowledgeServerName | null;
+  onSave: (action: AgentBuilderAction) => void;
+  isOpen: boolean;
+  onClose: () => void;
+  action?: AgentBuilderAction;
+}
+
+export function CapabilitiesConfigurationSheet({
+  capability,
+  onSave,
+  isOpen,
+  onClose,
+  action,
+}: CapabilitiesConfigurationSheetProps) {
+  // We store as state to control the timing to update the content for exit animation.
+  const [config, setConfig] = useState<CapabilityConfig | null>(null);
+
+  const handleClose = () => {
+    onClose();
+
+    // Wait until closing animation ends, otherwise exit animation won't work.
+    setTimeout(() => {
+      setConfig(null);
+    }, 200);
+  };
+
+  useEffect(() => {
+    if (isOpen) {
+      setConfig(capability ? CAPABILITY_CONFIGS[capability] : null);
+    }
+  }, [capability, isOpen]);
+
+  return (
+    <MultiPageSheet
+      open={isOpen}
+      onOpenChange={(open) => {
+        !open && handleClose();
+      }}
+    >
+      {config && (
+        <CapabilitiesConfigurationSheetContent
+          config={config}
+          onClose={handleClose}
+          action={action}
+          onSave={onSave}
+        />
+      )}
+    </MultiPageSheet>
+  );
+}
+
+interface CapabilitiesConfigurationSheetContent {
+  config: CapabilityConfig;
+  action?: AgentBuilderAction;
+  onSave: (action: AgentBuilderAction) => void;
+  onClose: () => void;
+}
+
+// This component gets unmounted when config is null so no need to reset the state.
+function CapabilitiesConfigurationSheetContent({
+  action,
+  onSave,
+  config,
+  onClose,
+}: CapabilitiesConfigurationSheetContent) {
+  const { owner, supportedDataSourceViews } = useAgentBuilderContext();
+  const { spaces } = useSpacesContext();
+  const instructions = useWatch<AgentBuilderFormData, "instructions">({
+    name: "instructions",
+  });
+
+  const [currentPageId, setCurrentPageId] = useState<ConfigurationSheetPageId>(
+    CONFIGURATION_SHEET_PAGE_IDS.DATA_SOURCE_SELECTION
+  );
+  const [dataSourceConfigurations, setDataSourceConfigurations] =
+    useState<DataSourceViewSelectionConfigurations>(() =>
+      getDataSourceConfigurations(action)
+    );
+
+  const form = useForm<CapabilityFormData>({
+    resolver: zodResolver(capabilityFormSchema),
+    defaultValues: {
+      description: action?.description ?? "",
+      timeFrame: getTimeFrame(action),
+      jsonSchema: getJsonSchema(action),
+    },
+  });
+
+  const { watch, setValue, getValues, formState } = form;
+  const [description, timeFrame, jsonSchema] = watch([
+    "description",
+    "timeFrame",
+    "jsonSchema",
+  ]);
+
+  const hasDataSources = hasDataSourceSelections(dataSourceConfigurations);
+
+  const handleSave = () => {
```

Comment:
I'm having difficult time how to manage this, but I assume I should get formData onSubmit instead of getting it from `getValues` or is it okay?